File: 1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/1379.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def dfs(self, root:TreeNode, target: TreeNode) -> TreeNode:
        if root is None:
            return None

        if root.val == target.val:
            return root

        return self.dfs(root.left, target) or self.dfs(root.right, target)

    def getTargetCopy(self, original: TreeNode, cloned: TreeNode, target: TreeNode) -> TreeNode:
        return self.dfs(cloned, target)

File: 1379-find-a-corresponding-node-of-a-binary-tree-in-a-clone-of-that-tree/1379.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
private:
    TreeNode* dfs(TreeNode *cloned, TreeNode *target) {
        if (cloned->val == target->val)
            return cloned;

        TreeNode* left = nullptr;
        TreeNode* right = nullptr;
        if (cloned->left)
            left = dfs(cloned->left, target);
        if (cloned->right)
            right = dfs(cloned->right, target);

        if (left)
            return left;
        else
            return right;
    }
public:
    TreeNode* getTargetCopy(TreeNode* original, TreeNode* cloned, TreeNode* target) {
        return dfs(cloned, target);
    }
};
